### PR TITLE
feat: gpu autotune

### PR DESCRIPTION
**Pull request by sshaman1101**
This PR adds GPU vendor detection and automatic tuning according to detected type.

### CODE REVIEW

File: insonmnia/miner/config.go [hunk 1]
```diff
@@ -82,8 +79,14 @@ func (c *config) PublicIPs() []string {
 	return c.PublicIPsConfig
 }
 
-func (c *config) GPU() *gpu.Config {
-	return c.GPUConfig
+func (c *config) GPU() pb.GPUVendorType {
+	t := strings.ToUpper(c.GPUConfig)
+	v, ok := pb.GPUVendorType_value[t]
```
Comment (Sokel) [inline insonmnia/miner/config.go]: it's may been more stylish
```
	if v, ok := pb.GPUVendorType_value[t]; ok {
		return pb.GPUVendorType(v)
	}
```

File: insonmnia/miner/config.go [hunk 2]
```diff
@@ -82,8 +79,14 @@ func (c *config) PublicIPs() []string {
 	return c.PublicIPsConfig
 }
 
-func (c *config) GPU() *gpu.Config {
-	return c.GPUConfig
+func (c *config) GPU() pb.GPUVendorType {
+	t := strings.ToUpper(c.GPUConfig)
```
Comment (Sokel) [inline insonmnia/miner/config.go]: this string should be trimed

File: insonmnia/miner/config_test.go [hunk 3]
```diff
@@ -54,5 +53,29 @@ locator:
...
+func TestGpuConfigTypes(t *testing.T) {
+	tests := []struct {
+		in  string
+		out sonm.GPUVendorType
+	}{
+		{in: "nvidia", out: sonm.GPUVendorType_NVIDIA},
+		{in: "Nvidia", out: sonm.GPUVendorType_NVIDIA},
+		{in: "NVIDIA", out: sonm.GPUVendorType_NVIDIA},
+
+		{in: "radeon", out: sonm.GPUVendorType_RADEON},
```
Comment (Sokel) [inline insonmnia/miner/config_test.go]: please cover behavior for case "AMD"

File: worker.yaml [hunk 4]
```diff
@@ -43,10 +43,9 @@ logging:
 # Ignored if firewall settings are not null.
 # public_ip_addrs: ["46.148.198.134"]
 
-# GPUConfig:
-  # type: "nvidiadocker"
-  # by default nvidia-docker-plugin works on that endpoint
-  # args: { nvidiadockerdriver: "localhost:3476" }
+# Preferred type for GPU tuner, may be "nvidia" or "radeon".
+# Any other value will disable auto GPU tuning.
+GPUConfig: "any"
```
Comment (Sokel) [inline worker.yaml]: just temporary conclusion - for nvidia+radeon system its will be crashed or not properly working

File: insonmnia/hardware/gpu/device_test.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+	"testing"
+
+	"fmt"
+
+	"github.com/sonm-io/core/proto"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+var deviceIdsTable = []struct {
```
Comment (Sokel) [inline insonmnia/hardware/gpu/device_test.go]: sorry guys, but i think it worse solution of all to do this.
its not be determined properly only by yours list of vendor number.
One of solution which can i offer you - use [opencl library](https://github.com/KhronosGroup/OpenCL-Headers/blob/master/opencl11/CL/cl.h#L473)  
[ps](https://www.kryptonitelock.com/content/dam/kryt-us/menu-svgs/bicycle.svg)

**Comment by Sokel** [comment]
@3Hren assure me that if it will broken, we will fix it.

**Comment by 3Hren** [comment]
@Sokel 
image

**Pull request merged**